### DISCUSSION TITLE
Question: Does Elsa support multi-tenancy ?

### DISCUSSION THREAD

**Discussion by iyhammad** [elsa-workflows/elsa-core#140] [Q&A]
I would like to ask, does Elsa support multi-tenancy ? If yes, what's the best way to do this ?

**Comment by sfmskywalker** [comment]
For sure, but Elsa itself does not provide anything to that end. Instead, it would be up to your application hosting the workflow runtime to provide a multi-tenant workflow definition and workflow instance store.

For example, if you were to use Orchard Core Framework to build a multi-tenant solution, Elsa would automatically be multi-tenant de to the fact that the service container itself would be tenant-specific.

If you would use ABP framework, the principle would probably be similar.

And of course you could implement your own multi-tenant data store.

Hope this helps.

**Comment by iyhammad** [comment]
Thank You. I'm planning to start using it in few days.

**Comment by Andale-online** [comment]
Hey... Found a simple way: if you e.g. use mongodb you could change connection strings to the representative bucket for the tenant... Just an idea. You just need some logic on the dashboard side to look up the right basket.

**Comment by slarkerino** [comment]
How to implement elsa-core with abp using mutlitenant? Please Help

**Reply by sfmskywalker** [reply]
The multitenancy story hasn't been well defined yet for Elsa I'm afraid. Hopefully I get the opportunity after the Elsa 2 release to dig into multitenancy.

**Comment by craigfowler** [comment]
My 'day job' customer actually is integrating Elsa (1.4.x) into a multi-tenancy environment with Entity Framework (your current authenticated user claims determine which Entity Framework db you're connecting to) .  You have just reminded me that I have a contribs to make toward Elsa 2.0 which can make that a little easier!

That'll be better for us then, when we come to upgrade to Elsa 2 😄

**Reply by sfmskywalker** [reply]
Nice! I imagine we’d have some form of abstraction that provides the current tenant for a given request, and a number of implementations such as accessing the current user’s claims and others based on host headers etc. Not saying you’d have to implement all of them of course 😅

**Reply by craigfowler** [reply]
To me, a _logical first step_ in providing that functionality is simply to ensure that all of the persistence providers can natively get their "connection object" (db context, or whatever) at runtime using a service registration.

I noticed at least in Elsa 1.4 that the Entity Framework context object was being selected upfront. I recall we needed to write our own version of a service collection registration extension method in order to get what we wanted.

Some of that work might already be done even.

With that alone, the consumer of Elsa can integrate into their own definition/meaning of multi-tenancy just by using DI. I've added #748 to at least do that (or check it's already been done) for Entity Framework.
